### PR TITLE
Make InMemorySpanRecorder public; use sampling.priority to turn sampling on and off

### PR DESCRIPTION
**Pull request by yurishkuro**
Changes:
 # make InMemorySpanRecorder public (useful for testing instrumentation)
 # allow turning sampling on and off via `"sampling.priority"` tag
 # add more unit tests

cc @bensigelman

### CODE REVIEW

File: span.go [hunk 1]
```diff
@@ -135,7 +137,9 @@ func (s *spanImpl) FinishWithOptions(opts opentracing.FinishOptions) {
 	s.raw.Duration = duration
 
 	s.onFinish(s.raw)
-	s.tracer.options.Recorder.RecordSpan(s.raw)
+	if s.raw.Sampled {
+		s.tracer.options.Recorder.RecordSpan(s.raw)
+	}
```
Comment (yurishkuro) [inline span.go]: Above is the most controversial change in this PR. Another alternative is to keep the old code, and add a `GetSampledSpans()` method to the `InMemorySpanRecorder`.

File: span.go [hunk 2]
```diff
@@ -74,8 +74,10 @@ func (s *spanImpl) SetTag(key string, value interface{}) opentracing.Span {
 	s.Lock()
 	defer s.Unlock()
 	if key == string(ext.SamplingPriority) {
-		s.raw.Sampled = true
-		return s
+		if v, ok := value.(uint16); ok {
+			s.raw.Sampled = v != 0
+			return s
+		}
```
Comment (yurishkuro) [inline span.go]: this is another behavior change, it allows forcing sampling both on and off via SamplingPriority tag.

Comment (bensigelman) [inline span.go]: sgtm

File: span.go (hunk 1 above)
Comment (bensigelman) [inline span.go]: I'd rather add `GetSampledSpans()` to IMSR.

File: tracer.go [hunk 3]
```diff
@@ -185,7 +185,11 @@ func (t *tracerImpl) startSpanInternal(
 	tags opentracing.Tags,
 ) opentracing.Span {
 	sp.tracer = t
-	sp.event = t.options.NewSpanEventListener()
```
Comment (bensigelman) [inline tracer.go]: was this broken before?

File: span.go (hunk 1 above)
Comment (bensigelman) [inline span.go]: (and then there's `TrimUnsampledSpans` to consider as well...)

File: tracer.go (hunk 3 above)
Comment (yurishkuro) [inline tracer.go]: it was forcing users to always provide an even listener if they were using `NewWithOptions()`. This interface does not feel like a required part of the tracer.

**Comment by yurishkuro** [comment]
@bensigelman ready for final review (diff is smaller now)

File: tracer.go (hunk 3 above)
Comment (bensigelman) [inline tracer.go]: +1

**Comment by bensigelman** [comment]
:ship:

**Pull request merged**